### ISSUE TITLE
The shellcheck external source(s) are set incorrectly (especially, when a script is in a subdirectory)

### ISSUE THREAD

**Issue by 64kramsystem** [bash-lsp/bash-language-server#624]
## Summary

When a script is in a subdirectory of a root, the plugin sends incorrect `--source` parameters, causing `SC1091` when there are sourced files.

Possibly related to #464.

## Details

I have multiple files that source a script in the same subdirectory; Shellcheck needs some related commandline options to be set, otherwise, it raises an `SC1091`.

Outside the VSC context, the typical solution is to pass `-x -P SCRIPTDIR` (`--external-sources --source-path=SCRIPTDIR`).

In the context of Bash IDE, I've captured the options passed, and I've observed two things:

1. the shell file passed is `-`; I assume it's because the content is piped;
2. for the previous reason, Bash IDE passes multiple directory paths as sources; specially, it passes the workspace roots.

Now, passing the roots is wrong for two reasons:

1. there could be aliasing, and the incorrect sourced file could be followed
2. if a script is in a subdirectory, the sources are effectively wrong, since they will not include the subdirectory

I'm not familiar the VSC extensions writing, however, the correct solution is to pass, for each script, its enclosing directory, instead of all the workspace roots.

## Example

```
+- mydir1
+- mydir2
   + subdir
      + myscript.sh
```

Bash IDE currently sends something like:

```sh
--shell=bash --format=json1 --external-sources --source-path=mydir1 --source-path=mydir2 -`
```

Instead, it should send:

```sh
--shell=bash --format=json1 --external-sources --source-path=/path/to/mdir2/subdir -`
```

**Comment by skovhus** [comment]
> I'm not familiar the VSC extensions writing, however, the correct solution is to pass, for each script, its enclosing directory, instead of all the workspace roots.

The source path actually depends on where the script is being executed from. I'm unsure how we should handle this in a generic way. Maybe pass in multiple `--source-path` with both workspace root and the sub folders?

Example: the scripts here https://github.com/bash-lsp/bash-language-server/blob/main/scripts/release-server.sh are executed from the root, so setting the workspace root as the `--source-path` is correct.

**Comment by 64kramsystem** [comment]
> > I'm not familiar the VSC extensions writing, however, the correct solution is to pass, for each script, its enclosing directory, instead of all the workspace roots.
> 
> The source path actually depends on where the script is being executed from. I'm unsure how we should handle this in a generic way. Maybe pass in multiple `--source-path` with both workspace root and the sub folders?
> 
> Example: the scripts here https://github.com/bash-lsp/bash-language-server/blob/main/scripts/release-server.sh are executed from the root, so setting the workspace root as the `--source-path` is correct.

Thanks for your reply :smile:

Now, I'm not familiar with TS/VSC development, however, I had a look at the source code, and based on my understanding, linting is performed by the [runShellCheckMethod()](https://github.com/bash-lsp/bash-language-server/blob/main/server/src/shellcheck/index.ts#L60). This method receives a TextDocument which, based on the [M$ documentation](https://code.visualstudio.com/api/references/vscode-api#TextDocument), has a `fileName` field.

Wouldn't it work to pass the directory name of the document (that is, `fileName` stripped of the filename) as `--source-path`? This simulates the `-P SCRIPTDIR` behavior of Shellcheck as closely as possible, and I think it covers file sourcing without needing to add all the roots.

Of course, for new documents, there isn't really any meaningful default until the document is saved (in this case, I guess that the document will have a blank `fileName`).

**Comment by skovhus** [comment]
> Wouldn't it work to pass the directory name of the document (that is, fileName stripped of the filename) as --source-path? 

We would still need to pass in the workspace root in case the user is executing the script from the root folder. 

But I believe this should be fixed now in bash-language-server@4.1.2 and in the latest vscode extension 1.23.0.

Can you test it out and let me know if it works for you? I appreciate the detailed error report here – keep them coming. ;)

**Comment by 64kramsystem** [comment]
Problem fixed! Thanks for your work :smile:
